### DISCUSSION TITLE
Data List: How to set value programmatically?

### DISCUSSION THREAD

**Discussion by Sven883** [leekelleher/umbraco-contentment#236] [Discussion]
### Which Contentment version are you using?

V1.4.4

### Which Umbraco version are you using? For example: 8.14.1 - don't just write v8

v8.6.1

### Bug summary

I've got 2 custom propertyeditors made:

1. SliderFiltersDataSource
```
public class SliderFiltersDataSource : IDataListSource, IDataListSourceValueConverter
    {
        public Dictionary<string, object> DefaultValues => new Dictionary<string, object>();

        public IEnumerable<ConfigurationField> Fields => Enumerable.Empty<ConfigurationField>();

        public OverlaySize OverlaySize => OverlaySize.Large;

        public string Name => "Slider Filters";

        public string Description => "Data source for all available slider filters.";

        public string Icon => "icon-navigation-horizontal";

        public IEnumerable<DataListItem> GetItems(Dictionary<string, object> config)
        {
            var sliderFilters = Current.UmbracoContext.Content.GetByContentType(SliderFilterTaxonomy.GetModelContentType()).OfType<SliderFilterTaxonomy>();

            if (sliderFilters == null || !sliderFilters.Any())
            {
                return new List<DataListItem>();
            }

            return sliderFilters.Select(x => new DataListItem
            {
                Name = x.Name,
                Icon = "icon-navigation-horizontal",
                Value = JsonConvert.SerializeObject(new Models.Custom.Taxonomy 
                { 
                    Id = x.Id,
                    Name = x.Name,
                    ContentTypeAlias = SliderFilterTaxonomy.ModelTypeAlias,
                    Key = x.Key,
                    Url = x.Url,
                    Parent = new Models.Custom.Taxonomy
                    {
                        Id = x.Parent.Id,
                        Name = x.Parent.Name,
                        ContentTypeAlias = SliderFiltersContainer.ModelTypeAlias,
                        Key = x.Parent.Key,
                        Url = x.Parent.Url
                    }
                })
            });
        }

        public Type GetValueType(Dictionary<string, object> config)
        {
            return typeof(Models.Custom.Taxonomy);
        }

        public object ConvertValue(Type type, string value)
        {
            return JsonConvert.DeserializeObject<Models.Custom.Taxonomy>(value);
        }
    }
```

2. CheckboxFiltersDataSource

```
    public class CheckboxFiltersDataSource : IDataListSource, IDataListSourceValueConverter
    {
        public Dictionary<string, object> DefaultValues => new Dictionary<string, object>();

        public IEnumerable<ConfigurationField> Fields => Enumerable.Empty<ConfigurationField>();

        public OverlaySize OverlaySize => OverlaySize.Large;

        public string Name => "Checkbox Filters";

        public string Description => "Data source for all available checkbox filters.";

        public string Icon => "icon-checkbox";

        public IEnumerable<DataListItem> GetItems(Dictionary<string, object> config)
        {
            var checkboxFilters = Current.UmbracoContext.Content.GetByContentType(CheckboxFilterTaxonomy.GetModelContentType()).OfType<CheckboxFilterTaxonomy>();

            if (checkboxFilters == null || !checkboxFilters.Any())
            {
                return new List<DataListItem>();
            }

            return checkboxFilters.Select(x => new DataListItem
            {
                Name = x.Name,
                Icon = "icon-checkbox",
                Value = JsonConvert.SerializeObject(new Models.Custom.Taxonomy
                {
                    Id = x.Id,
                    Name = x.Name,
                    ContentTypeAlias = CheckboxFilterTaxonomy.ModelTypeAlias,
                    Key = x.Key,
                    Url = x.Url,
                    Parent = new Models.Custom.Taxonomy
                    {
                        Id = x.Parent.Id,
                        Name = x.Parent.Name,
                        ContentTypeAlias = CheckboxFiltersContainer.ModelTypeAlias,
                        Key = x.Parent.Key,
                        Url = x.Parent.Url
                    }
                }),
            });
        }

        public Type GetValueType(Dictionary<string, object> config)
        {
            return typeof(Models.Custom.Taxonomy);
        }

        public object ConvertValue(Type type, string value)
        {
            return JsonConvert.DeserializeObject<Models.Custom.Taxonomy>(value);
        }
    }
}
```

I created a custom app_plugin so products with old filtering are parsed to new filtering with above new slider & checkboxes.
Problem i'm having is that either way how I try to set the value, the value is always empty in the backoffice: 

1. checkboxFilters property = contentment datalist
I've tried setting the value like this: 

```
var newCheckboxValues = new List<Models.Custom.Taxonomy>();
newCheckboxValues.Add(new Models.Custom.Taxonomy 
                    { 
                        ContentTypeAlias = CheckboxFilterTaxonomy.ModelTypeAlias,
                        Id = checkboxFilter.Id,
                        Key = checkboxFilter.Key,
                        Name = checkboxFilter.Name,
                        Url = checkboxFilter.Url,
                        Parent = new Models.Custom.Taxonomy
                        {
                            Id = checkboxFilter.Parent.Id,
                            Name = checkboxFilter.Parent.Name,
                            ContentTypeAlias = CheckboxFiltersContainer.ModelTypeAlias,
                            Key = checkboxFilter.Parent.Key,
                            Url = checkboxFilter.Parent.Url
                        }
                    });
productItem.SetValue("checkboxFilters", JsonConvert.SerializeObject(newCheckboxValues));
```

But that doesn't work. Ive also tried returning a array of only the names, but same result.

3. sliderFilters property= Nested content with properties: type (contentment datalist max 1 item), min (numeric), max (numeric)
For the sliderFilter i've created a custom model: 
```
public class NcSliderFilter
    {
        public Guid Key { get; set; }

        public string NcContentTypeAlias { get; set; }

        public int Min { get; set; }

        public int Max { get; set; }

        public Models.Custom.Taxonomy Type { get; set; }
    }
```

```
var nestedContentValue = new List<NcSliderFilter>();
var type = new Models.Custom.Taxonomy
                {
                    Id = sliderFilter.Id,
                    Name = sliderFilter.Name,
                    ContentTypeAlias = SliderFilterTaxonomy.ModelTypeAlias,
                    Key = sliderFilter.Key,
                    Url = sliderFilter.Url,
                    Parent = new Models.Custom.Taxonomy
                    {
                        Id = sliderFilter.Parent.Id,
                        Name = sliderFilter.Parent.Name,
                        ContentTypeAlias = SliderFiltersContainer.ModelTypeAlias,
                        Key = sliderFilter.Parent.Key,
                        Url = sliderFilter.Parent.Url
                    }
                };

                var ncContent = new NcSliderFilter
                {
                    Key = Guid.NewGuid(),
                    Max = product.Value<int>(availableSliderFilter),
                    Min = availableSliderFilter.Equals("liftingHeightMm") ? product.Value<int>("liftingHeightMinimumMm") : 0,
                    NcContentTypeAlias = SliderFilter.ModelTypeAlias,
                    Type = type
                };


productItem.SetValue("sliderFilters", JsonConvert.SerializeObject(nestedContentValue));
```

I've tried adding the value with/without serializing, but no results.
Any idea would be greatly appreciated!
Thanks in advance

**Comment by Sven883** [comment]
Hey @leekelleher, i've gottan a bit further since my post. 

this is the json data that I need: 
`"[\r\n  \"{\\\"Id\\\":5469,\\\"Key\\\":\\\"cdb2b8c2-3ef0-4a22-8d8c-950cedfddb03\\\",\\\"Name\\\":\\\"Meeloper\\\",\\\"Url\\\":\\\"/checkbox-filter-container-nl-1-1-1-1/extra/meeloper/\\\",\\\"ContentTypeAlias\\\":\\\"checkboxFilterTaxonomy\\\",\\\"Parent\\\":{\\\"Id\\\":5468,\\\"Key\\\":\\\"4dee4dcf-26b6-41e6-b213-51617d582f32\\\",\\\"Name\\\":\\\"Extra\\\",\\\"Url\\\":\\\"/checkbox-filter-container-nl-1-1-1-1/extra/\\\",\\\"ContentTypeAlias\\\":\\\"checkboxFiltersContainer\\\",\\\"Parent\\\":null}}\",\r\n  \"{\\\"Id\\\":5474,\\\"Key\\\":\\\"5a1441bb-f2b5-440a-b497-8ef2c4e2e94e\\\",\\\"Name\\\":\\\"Steekwagens\\\",\\\"Url\\\":\\\"/checkbox-filter-container-nl-1-1-1-1/type-machine/steekwagens/\\\",\\\"ContentTypeAlias\\\":\\\"checkboxFilterTaxonomy\\\",\\\"Parent\\\":{\\\"Id\\\":5473,\\\"Key\\\":\\\"70c6d8a6-6415-4935-9192-4be95b52d904\\\",\\\"Name\\\":\\\"Type machine\\\",\\\"Url\\\":\\\"/checkbox-filter-container-nl-1-1-1-1/type-machine/\\\",\\\"ContentTypeAlias\\\":\\\"checkboxFiltersContainer\\\",\\\"Parent\\\":null}}\"\r\n]"`

When copy pasting this in:
`productItem.SetValue("checkboxFilters", jsonDataAboveCopyPastedAsString);`

It works.

But weirdly enough it doesn't work when doing this: 
```
var serializedObject = JsonConvert.SerializeObject(newCheckboxValues, Formatting.Indented);
productItem.SetValue("checkboxFilters",serializedObject );
```

Any idea?

**Comment by leekelleher** [comment]
@Sven883 Thanks for posting this. There is quite a bit to unpack from your code snippets.  My general approach would be to try to simplify the value of the data-source items.

From your code snippet examples, you are serializing a new `Models.Custom.Taxonomy` object instance to a JSON-string, which contains references to the content node's name and info about parent's node - as in the `Value` is capturing a snapshot of the node at that moment in time. e.g. What would happen if a category/filter was renamed? or if it was relocated to under a different parent category/filter? In that scenario, that would break the corresponding `Value`, it'd be out-of-sync.

When I've been working with content node data-sources for Data List, I try to use the node's `.Key` (`Guid`), or the UDI (e.g. `umb://document/<guid>`), then later in the `ConvertValue()` method, I'd use the UDI to query the content cache.

Here's an example of how I do that here: <https://github.com/leekelleher/umbraco-contentment/blob/1.4.4/src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/UmbracoContentXPathDataListSource.cs#L121-L123>

---

However, if you are determined to have the serialized JSON-string as the `.Value`, then when you want to set the property's value programmatically, you must match the value literally. Meaning that when you serialize it, it must use the same settings, e.g. `Formatting.Indented`, (personally, I wouldn't indent it), and all the object's inner values must match.  The JSON-strings must be literally equal, otherwise the Data List editor wont be able to match them up.

I hope this helps?
